src/audio/drc/drc.c

```diff
@@ -139,353 +141,234 @@ static int drc_setup(struct drc_comp_data *cd, uint16_t channels, uint32_t rate)
+	audio_stream_init_alignment_constants(1, 1, sink);
 }
 
-/* copy and process stream data from source to sink buffers */
-static int drc_copy(struct comp_dev *dev)
+static int drc_process(struct processing_module *mod,
+		       struct input_stream_buffer *input_buffers,
+		       int num_input_buffers,
+		       struct output_stream_buffer *output_buffers,
+		       int num_output_buffers)
 {
-	struct comp_copy_limits cl;
-	struct drc_comp_data *cd = comp_get_drvdata(dev);
-	struct comp_buffer *sourceb, *sinkb;
-	struct comp_buffer __sparse_cache *source_c, *sink_c;
-	int ret = 0;
-
-	comp_dbg(dev, "drc_copy()");
+	struct drc_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct audio_stream __sparse_cache *source = input_buffers[0].data;
+	struct audio_stream __sparse_cache *sink = output_buffers[0].data;
+	int frames = input_buffers[0].size;
+	int ret;
 
-	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer,
-				  sink_list);
-	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer,
-				source_list);
-
-	source_c = buffer_acquire(sourceb);
-	sink_c = buffer_acquire(sinkb);
+	comp_dbg(dev, "drc_process()");
 
 	/* Check for changed configuration */
 	if (comp_is_new_data_blob_available(cd->model_handler)) {
 		cd->config = comp_get_data_blob(cd->model_handler, NULL, NULL);
-		ret = drc_setup(cd, audio_stream_get_channels(&source_c->stream),
-				audio_stream_get_rate(&source_c->stream));
+		ret = drc_setup(cd, audio_stream_get_channels(sink),
+				audio_stream_get_rate(sink));
 		if (ret < 0) {
 			comp_err(dev, "drc_copy(), failed DRC setup");
-			goto out;
+			return ret;
 		}
 	}
 
-	/* Get source, sink, number of frames etc. to process. */
-	comp_get_copy_limits(source_c, sink_c, &cl);
+	cd->drc_func(mod, source, sink, frames);
+
+	/* calc new free and available */
+	module_update_buffer_position(&input_buffers[0], &output_buffers[0], frames);
+	return 0;
+}
 
-	/* Run DRC function */
-	drc_process(dev, source_c, sink_c, cl.frames, cl.source_bytes,
-		    cl.sink_bytes);
+#if CONFIG_IPC_MAJOR_4
+static void drc_params(struct processing_module *mod)
+{
+	struct sof_ipc_stream_params *params = mod->stream_params;
+	struct comp_buffer __sparse_cache *sink_c, *source_c;
+	struct comp_buffer *sinkb, *sourceb;
+	struct comp_dev *dev = mod->dev;
 
-out:
+	comp_dbg(dev, "drc_params()");
+
+	ipc4_base_module_cfg_to_stream_params(&mod->priv.cfg.base_cfg, params);
+	component_set_nearest_period_frames(dev, params->rate);
+
+	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+	sink_c = buffer_acquire(sinkb);
+	ipc4_update_buffer_format(sink_c, &mod->priv.cfg.base_cfg.audio_fmt);
 	buffer_release(sink_c);
-	buffer_release(source_c);
 
-	return ret;
+	sourceb = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	source_c = buffer_acquire(sourceb);
+	ipc4_update_buffer_format(source_c, &mod->priv.cfg.base_cfg.audio_fmt);
```

Comment:
oh gosh, grepping for `base_cfg.audio_fmt`  is all over the place. Mind adding a macro for `mod` to `audio_fmt`? Can be done in another PR.

Reply:
ack, I think there will be a lot of low hanging fruit for convenience macros to take care of cleanup and readability after all the modules are converted.